brave/brave-core
# feat(wallet): Add support for Retry, Cancel & Speed Up to transactions on iOS

- Add `Retry Transaction` button to transaction details & long press context menu for failed ethereum transactions
- Add `Cancel Transaction` & `Speed Up` buttons to transaction details & long press context menu for failed ethereum transactions

Resolves https://github.com/brave/brave-browser/issues/36627

## Submitter Checklist:

  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`

## Reviewer Checklist:


## After-merge Checklist:

  changes has landed on

## Test Plan:

Cancel Transaction / Speed Up Transaction:
1. Create an EVM transaction, do not approve it yet. 
   - Tip: Can yse Goerli network now that the network is deprecated. Transaction will stay in `Submitted` state allowing cancel / retry buttons to appear.
2. Dismiss pending transaction modal, switch to activity tab.
3. Open pending transaction and approve, dismiss confirmation quick
4. Tap and hold on Activity tab row & verify `Cancel Transaction` and `Speed Up Transaction` rows are shown in context menu
5. Tap the transaction row to open transaction details
6. Verify `Cancel Transaction` and `Speed Up` are shown. 
7. Tapping one of these buttons will create a new unapproved transaction.

Retry Transaction:
1. Create an EVM transaction.
2. In the transaction approval screen, edit nonce to below last transaction nonce
    - I just used 1 on an account that had >1 transactions previously on Sepolia
3. Approve the transaction. We need transaction to fail so setting low gas fee can help here.
4. Visit Activity tab and tap & hold on the failed transaction (indicated by red dot)
5. Verify `Retry Transaction` is shown in context menu
6. Tap the transaction row to open transaction details
7. Verify `Retry Transaction` is shown
8. Tap on the button to create a new unapproved transaction. 
    - You will have to manually change nonce value on this transaction as it will be re-created with same value used in step 2.

![Cancel Speed Up Tx Details](https://github.com/brave/brave-core/assets/5314553/26329589-ac04-4a62-b437-da61c11eb686) | ![Cancel Speed Up Context Menu](https://github.com/brave/brave-core/assets/5314553/b677b29b-1d4e-49b9-8c40-2a093a5c93cf)
--|--
![Retry Tx Details](https://github.com/brave/brave-core/assets/5314553/cf1643c4-0404-4108-aec3-8e471508a362) | ![Retry Context Menu](https://github.com/brave/brave-core/assets/5314553/58ed6ede-1e04-4e0d-a018-50ee64feed73)
